### PR TITLE
[torch.Package/TorchScript] ScriptModuleSerializer add unified format

### PR DESCRIPTION
**Pull request by lobstergrindset**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57888 [PyTorch Edge] bytecode version bump to v5 and enable share constant table
* #58182 [PyTorch Edge] share tensors in mobile with new api
* #57678 [torch.package/TorchScript] flag to gate allowance of TS serializaiton in torch.package
* #54894 [torch.Package/TorchScript] TS into torch.Package test cases
* #57573 [torch.package/TorchScript] logic to enable sharing of tensors on load
* #54893 [torch.Package/TorchScript] torch.Package python logic to save TorchScript
* **#56299 [torch.Package/TorchScript] ScriptModuleSerializer add unified format**
* #55958 [torch.Package/TorchScript] refactor ScriptModuleSerializer Exporter
* #54891 [torch.Package/TorchScript] TS serialization importer to handle unified format

Differential Revision: [D27832545](https://our.internmc.facebook.com/intern/diff/D27832545)

### CODE REVIEW

**Review by suo — APPROVED** [review]
Seems fine overall, but I think given the risk of making these changes to the serialization format we should do a round of manual testing before landing.

The expected behavior is that all existing usages of ScriptModuleSerializer should be exactly the same as before this stack. So I propose we do the following:
1. instrument the entire test suite so that every ScriptModule is saved to disk (using the emitModuleHook) at some deterministic location
2. Run the test suite on master and on this stack. We probably need to fix the rng and run the jit tests sequentially to get determinism
3. Do a checksum on the saved files to ensure they are exactly the same.

This would give us really high confidence that this stack is safe to land as is. If we run into hurdles (weird nondeterminism preventing checksums from being equal), we can hand select some representative models and spot check them, but it would be nice to do the whole test suite if possible.

**Comment by Lilyjjo** [comment]
Changes since last review:
- ScriptModuleSerializer is now using unique_id() for torchscript id 
- Tensors are shared between eager and ScriptModules
- ScriptModuleSerializer is using storage's _cdata pointer in naming scheme for tensors
- ScriptModuleSerializer will check PyTorchStreamWriter to see if storage is already serailized

File: torch/csrc/jit/serialization/export_module.cpp [hunk 1]
```diff
@@ -403,10 +410,14 @@ void ScriptModuleSerializer::serialize(
 
 void ScriptModuleSerializer::writeArchive(
     const IValue& value,
-    const std::string& archive_name) {
+    const std::string& archive_name,
+    const std::string& archive_dir,
+    const std::string& tensor_dir,
+    bool tensor_cdata_naming_scheme) {
```
Comment (suo) [inline torch/csrc/jit/serialization/export_module.cpp:416]: I don't love that we are forking the tensor naming scheme here but I guess following the principle of "don't change existing serialization…yet" it is necessary. We definitely need to follow up to unify all this though.

File: torch/csrc/jit/serialization/export_module.cpp [hunk 2]
```diff
@@ -415,18 +426,44 @@ void ScriptModuleSerializer::writeArchive(
...
+  const std::vector<std::string>& pre_serialized_files =
+      writer_.getAllWrittenRecords();
+
   for (const auto& td : data_pickle.tensorData()) {
     WriteableTensorData writable_td = getWriteableTensorData(td);
-    std::string fname = prefix + c10::to_string(i++);
-    writer_.writeRecord(fname, writable_td.data(), writable_td.sizeInBytes());
+    std::string file_name = tensor_dir + tensor_names[i++];
+    if (tensor_cdata_naming_scheme &&
+        std::find(
```
Comment (suo) [inline torch/csrc/jit/serialization/export_module.cpp:455]: maybe build an unordered_set so that we don't have to do a linear search for every tensor?

**Review by suo — APPROVED** [review]

File: torch/csrc/jit/serialization/export_module.cpp (hunk 2 above)
Comment (Lilyjjo) [reply torch/csrc/jit/serialization/export_module.cpp:455]: I was thinking that the tradeoff of copying the vector into an unordered set would be more costly than just doing the linear search. Right now the `writer_.getAllWrittenRecords()` returns a reference to its contents and isn't doing any copying or duplication (I think that's how cpp works?). I'm not sure if it's more expensive to do a lot of string comparisons or if it's more expensive to copy all of the string up front to do less comparisons

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832545).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832545).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832545).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832545).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832545).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832545).

File: torch/csrc/jit/serialization/pickler.cpp [hunk 3]
```diff
@@ -306,7 +306,13 @@ void Pickler::pushStorageOfTensor(const at::Tensor& tensor) {
   } else {
     root_key = c10::to_string(tensor_data_.size());
   }
+
+  // change will only be triggered on non-mobile logic
+  if (get_tensor_id_ != nullptr) {
+    root_key = get_tensor_id_(tensor);
+  }
```
Comment (cccclai) [inline torch/csrc/jit/serialization/pickler.cpp:313]: Move to `else` statement, after line 308

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832545).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832545).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832545).

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27832545).

**Pull request closed without merging**